edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+        void TriggerReconnect(object sender, EventArgs e)
+        {
+            Task.Run(async () =>
+            {
+                var client = default(MqttClient);
+
+                lock (this.guard)
+                {
+                    if (!this.mqttClient.HasValue)
+                    {
+                        return;
+                    }
+
+                    client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to use"));
+                }
+
+                // don't trigger it to ourselves while we are trying
+                client.ConnectionClosed -= this.TriggerReconnect;
+
+                var isConnected = false;
+
+                while (!isConnected)
+                {
+                    await Task.Delay(ReconnectDelayMs);
+
+                    lock (this.guard)
+                    {
+                        // seems Disconnect has been called since.
+                        if (!this.mqttClient.HasValue)
+                        {
+                            return;
+                        }
+
+                        client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to use"));
+                    }
+
+                    isConnected = await TryConnectAsync(client, this.components.Consumers, this.systemComponentIdProvider.EdgeHubBridgeId);
+                }
+
+                client.ConnectionClosed += this.TriggerReconnect;
+            });
+        }
+
+        Task StartForwardingLoop()
+        {
+            var loopTask = Task.Run(
+                                async () =>
+                                {
+                                    Events.ForwardingLoopStarted();
+                                    while (await this.publications.Expect(ChannelIsBroken).Reader.WaitToReadAsync())
+                                    {
+                                        var publishInfo = default(MqttPublishInfo);
+
+                                        try
+                                        {
+                                            publishInfo = await this.publications.Expect(ChannelIsBroken).Reader.ReadAsync();
+                                        }
+                                        catch (Exception e)
+                                        {
+                                            Events.FailedToForward(e);
+                                            continue;
+                                        }
+
+                                        foreach (var consumer in this.components.Consumers)
+                                        {
+                                            try
+                                            {
+                                                var accepted = await consumer.HandleAsync(publishInfo);
+                                                Events.MessageForwarded(consumer.GetType().Name, accepted, publishInfo.Topic, publishInfo.Payload.Length);
```

Comment:
Won't this flood your logs?

Reply:
Why not break the loop if a message is accepted

Reply:
I can change it to list only those accepted - or if it was not accepted by anything

Reply:
originally subscription changes were messages multiple consumers processed. Factoring that out into a separate class, maybe there is no more such message, so the loop could be broken. I need to check that